README.md
# fl_pin_code

[![pub package](https://img.shields.io/badge/build-success-green)](https://pub.dev/packages/fl_pin_code)
[![License: BSD-3-Clause](https://img.shields.io/badge/License-BSD--3--Clause-orange)](https://opensource.org/licenses/BSD-3-Clause)
![Dart language](https://img.shields.io/badge/Dart-100%25-blue)

# fl_pin_code

fl_pin_code helps to display Pin codes for Flutter applications.

## Feature 👇👇
* obscure support
* auto moving focus
* field shapes (box, underline)
* null-safety

## Installing 🔧
Install the latest version from [pub](https://pub.dartlang.org/packages/fl_pin_code).

# Installing
## Use this package as a library
### 1. Depend on it
Add this to your package's pubspec.yaml file:
```
dependencies:
  fl_pin_code: ^0.0.9
```
### 2. Install it
You can install packages from the command line:
with Flutter:
```
$ flutter packages get
```
Alternatively, your editor might support flutter packages get. Check the docs for your editor to learn more.

### 3. Import it
Now in your Dart code, you can use:
```
import 'package:fl_pin_code/fl_pin_code.dart';
```

### Usage
```
            PinCode(
                    numberOfFields: 5,
                    fieldWidth: 40.0,
                    style: TextStyle(color: Colors.black, fontSize: 15),
                    fieldStyle: PinCodeStyle.box),

            PinCode(
                  keyboardType: TextInputType.number,
                  isObscure: true,
                  numberOfFields: 5,
                  fieldWidth: 40.0,
                  style: TextStyle(color: Colors.black, fontSize: 15),
                  fieldStyle: PinCodeStyle.box,
                  onCompleted: (text) {
                    setState(() {
                      hash.pinHash = text;
                    });
                  },
            ),

            // this is taken from sample project.
            PinCode(
                numberOfFields: 5,
                fieldWidth: 40.0,
                style: TextStyle(color: Colors.black, fontSize: 15),
                fieldStyle: PinCodeStyle.box,
                onCompleted: (text) {
                  if (text.trim() == "11111") {
                    setState(() {
                      _passwordObscureText = false;
                    });
                  } else {
                    ScaffoldMessenger.of(context).showSnackBar(SnackBar(
                        content: Text('Please insert 11111 to unlock')));
                  }
                  Navigator.of(context).pop();
                },
                // onChanged: (_) {},
            )

```


CHANGELOG.md
## [0.0.9] - 16 March 2021

* Fix hidden crash when onChanged callback is null.
* Adding sample project to show how to use fl_pin_code.
* Adding focus nodes into widget.
* Minor bug fixes.

## [0.0.8] - 10 March 2021

* Obscuring text.
* Text input type for pin code widget.

## [0.0.7] - 10 March 2021

* Format code.

## [0.0.6] - 10 March 2021

* Remove redundant.

## [0.0.5] - 10 March 2021

* Update Project description.

## [0.0.4] - 10 March 2021

* Fix rendering fields.

## [0.0.2] - 10 March 2021

* Update Readme.

## [0.0.1] - 10 March 2021

* added pin code widget.


test/fl_pin_code_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:fl_pin_code/fl_pin_code.dart';

void main() {
  test('adds one to input values', () {
  });
}


lib/pin_code.dart
import 'package:fl_pin_code/styles.dart';
import 'package:flutter/material.dart';

class PinCode extends StatefulWidget {
  /// Number of Pin fields
  final int numberOfFields;

  /// Field width
  final double fieldWidth;

  /// Text style on each field
  final TextStyle style;

  /// Text keyboard type on fields
  final TextInputType keyboardType;

  /// Style on each field
  final PinCodeStyle fieldStyle;

  /// is obscure text?
  final bool isObscure;

  /// Called when pin changes its fields
  final ValueChanged<String>? onChanged;

  /// Called when pin is completed
  final ValueChanged<String>? onCompleted;

  PinCode(
      {Key? key,
      required this.numberOfFields,
      required this.fieldWidth,
      required this.style,
      required this.fieldStyle,
      this.keyboardType = TextInputType.number,
      this.isObscure = false,
      this.onChanged,
      this.onCompleted});

  @override
  _PinCodeState createState() => _PinCodeState();
}

class _PinCodeState extends State<PinCode> {
  late List<String> _pinCode;
  late List<TextEditingController?> _textControllers;
  late List<FocusNode?> _focusNodes;
  late List<Widget> _pinFields;

  @override
  void initState() {
    super.initState();

    _textControllers = List<TextEditingController?>.filled(
        widget.numberOfFields, null,
        growable: false);

    _focusNodes =
        List<FocusNode?>.filled(widget.numberOfFields, null, growable: false);

    _pinCode = List.generate(widget.numberOfFields, (index) => '');

    _pinFields = List.generate(
        widget.numberOfFields, (index) => _buildTextField(context, index));
  }

  @override
  void dispose() {
    super.dispose();
    _textControllers
        .forEach((TextEditingController? controller) => controller?.dispose());
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      width: MediaQuery.of(context).size.width,
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
        crossAxisAlignment: CrossAxisAlignment.center,
        children: _pinFields,
      ),
    );
  }

  Widget _buildTextField(BuildContext context, int index) {
    if (_focusNodes[index] == null) {
      _focusNodes[index] = new FocusNode();
    }

    if (_textControllers[index] == null) {
      _textControllers[index] = new TextEditingController();
    }

    return Container(
      width: widget.fieldWidth,
      child: TextField(
        textAlign: TextAlign.center,
        maxLength: 1,
        maxLines: 1,
        keyboardType: widget.keyboardType,
        style: widget.style,
        autofocus: false,
        obscureText: widget.isObscure,
        controller: _textControllers[index],
        focusNode: _focusNodes[index],
        decoration: InputDecoration(
          counterText: "",
          border: getInputDecoration(widget.fieldStyle),
        ),
        onChanged: (String value) {
          if (value.trim().isEmpty) {
            if (index == 0) return;

            _focusNodes[index]!.unfocus();
            _focusNodes[index - 1]!.requestFocus();
          }

          setState(() {
            _pinCode[index] = value;
          });

          if (value.trim().isNotEmpty) {
            _focusNodes[index]!.unfocus();
          }

          if (index + 1 != widget.numberOfFields && value.trim().isNotEmpty) {
            FocusScope.of(context).requestFocus(_focusNodes[index + 1]);
          }

          String currentPin = '';
          _pinCode.forEach((element) => currentPin += element);

          // if there is no any null value, then pin code is completed, automatically
          // call onCompleted callback
          if (!_pinCode.contains(null) &&
              !_pinCode.contains('') &&
              currentPin.length == widget.numberOfFields) {
            widget.onCompleted!(currentPin);
          }

          // call onChanged callback
          if (widget.onChanged != null) {
            widget.onChanged!(currentPin);
          }
        },
      ),
    );
  }

  InputBorder getInputDecoration(PinCodeStyle style) {
    switch (style) {
      case PinCodeStyle.box:
        return OutlineInputBorder(
            borderSide: BorderSide(width: widget.fieldWidth));
      case PinCodeStyle.underline:
      default:
        return UnderlineInputBorder(
            borderSide: BorderSide(width: widget.fieldWidth));
    }
  }
}


example/lib/main.dart
import 'package:fl_pin_code/fl_pin_code.dart';
import 'package:fl_pin_code/styles.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'fl_pin_code demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, this.title}) : super(key: key);

  final String? title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  bool _passwordObscureText = true;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title!),
      ),
      body: Padding(
        padding: EdgeInsets.all(10),
        child: Builder(
          builder: (context) {
            return Form(
              key: UniqueKey(),
              child: SingleChildScrollView(
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    SizedBox(
                      height: 20,
                    ),
                    TextFormField(
                      readOnly: true,
                      showCursor: false,
                      obscureText: _passwordObscureText,
                      initialValue: 'A_asdw[jd%_*klmdn',
                      decoration: InputDecoration(
                        suffixIcon: IconButton(
                          onPressed: () {
                            setState(() {
                              if (_passwordObscureText) {
                                showModalPinBottomSheet();
                              } else {
                                _passwordObscureText = true;
                              }
                            });
                          },
                          icon: Icon(
                            _passwordObscureText
                                ? Icons.visibility_off
                                : Icons.visibility,
                          ),
                          iconSize: 18.0,
                        ),
                      ),
                    ),
                    SizedBox(
                      height: 15,
                    ),
                  ],
                ),
              ),
            );
          },
        ),
      ),
    );
  }

  showModalPinBottomSheet() {
    showModalBottomSheet(
        isScrollControlled: true,
        shape: RoundedRectangleBorder(
            borderRadius: BorderRadius.only(
                topLeft: Radius.circular(25), topRight: Radius.circular(25))),
        context: context,
        builder: (_) {
          return Container(
            padding: EdgeInsets.only(bottom: 5),
            margin: EdgeInsets.symmetric(horizontal: 20),
            child: SingleChildScrollView(
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.center,
                children: [
                  SizedBox(
                    height: 10,
                  ),
                  PinCode(
                    numberOfFields: 5,
                    fieldWidth: 40.0,
                    style: TextStyle(color: Colors.black, fontSize: 15),
                    fieldStyle: PinCodeStyle.box,
                    onCompleted: (text) {
                      if (text.trim() == "11111") {
                        setState(() {
                          _passwordObscureText = false;
                        });
                      } else {
                        ScaffoldMessenger.of(context).showSnackBar(SnackBar(
                            content: Text('Please insert 11111 to unlock')));
                      }
                      Navigator.of(context).pop();
                    },
                    // onChanged: (_) {},
                  )
                ],
              ),
            ),
          );
        });
  }
}
